app/uk/gov/hmrc/selfservicetimetopay/controllers/CalculatorController.scala

```diff
@@ -18,94 +18,302 @@ package uk.gov.hmrc.selfservicetimetopay.controllers
 
 import play.api.data.Form
 import play.api.data.Forms._
+import play.api.data.validation.{Constraint, Invalid, Valid, ValidationError}
 import play.api.mvc._
 import uk.gov.hmrc.play.frontend.controller.FrontendController
-import uk.gov.hmrc.selfservicetimetopay.controllerVariables
+import uk.gov.hmrc.selfservicetimetopay.connectors.CalculatorConnector
 import uk.gov.hmrc.selfservicetimetopay.controllerVariables._
 import uk.gov.hmrc.selfservicetimetopay.models._
+import uk.gov.hmrc.selfservicetimetopay.util.JacksonMapper
 import views.html.selfservicetimetopay.calculator._
 
 import scala.concurrent.Future
 
+//@TODO replace session with keysotre
 object CalculatorController extends FrontendController {
 
   private def createAmountDueForm: Form[CalculatorAmountDue] = {
     Form(mapping(
       "amount" -> bigDecimal,
-      "dueByDay" -> number(min = 1, max = 31),
+      "dueByYear" -> number(min = 2000, max = 2100),
       "dueByMonth" -> nonEmptyText,
-      "dueByYear" -> number(min = 2000, max = 2100)
+      "dueByDay" -> number(min = 1, max = 31)
     )(CalculatorAmountDue.apply)(CalculatorAmountDue.unapply))
   }
 
-  private def createPaymentTodayForm: Form[CalculatorPaymentToday] = {
+
+  private def createPaymentTodayForm(totalDue:BigDecimal): Form[CalculatorPaymentToday] = {
     Form(mapping(
-      "amount" -> optional(bigDecimal)
+      "amount" -> optional(bigDecimal
+        .verifying("ssttp.calculator.form.payment_today.amount.less-than-owed", _ < totalDue)
+        .verifying("ssttp.calculator.form.payment_today.amount.nonnegitive", _.compare(BigDecimal("0")) >= 0)
+    )
     )(CalculatorPaymentToday.apply)(CalculatorPaymentToday.unapply))
   }
 
-  private def createDurationForm(min: Integer, max: Integer): Form[CalculatorDuration] = {
+  private def createDurationForm(min: Int, max: Int): Form[CalculatorDuration] = {
+    def required: Constraint[Int] = Constraint[Int]("constraint.required") { o =>
+      if (o == null) Invalid(ValidationError("ssttp.calculator.form.duration.months.required")) else Valid
+    }
+    def greaterThan: Constraint[Int] = Constraint[Int]("constraint.duration-more-than") { o =>
+      if (o != null && o < min) Invalid(ValidationError("ssttp.calculator.form.duration.months.greater-than", min)) else Valid
+    }
+    def lessThan: Constraint[Int] = Constraint[Int]("constraint.duration-more-than") { o =>
+      if (o != null && o > max) Invalid(ValidationError("ssttp.calculator.form.duration.months.less-than", max)) else Valid
+    }
     Form(mapping(
-      "months" -> number(min = min, max = max)
+      "months" -> number.verifying(required, greaterThan, lessThan)
     )(CalculatorDuration.apply)(CalculatorDuration.unapply))
   }
 
+  private def getKeystoreData(implicit request:Request[AnyContent]) : (Boolean, Option[CalculatorAmountsDue], Option[CalculatorPaymentToday], Option[CalculatorDuration], Option[Seq[CalculatorPaymentSchedule]]) = {
```

Comment:
Also use of Session/Keystore seems to be excessive I'd of thought that you'd want to keep it to a minimum in order to make the code maintainable, global mutable state is the source of many bugs. Bear in mind I don't really know the exact functionality here, but my assumption would be that the only one which is perhaps necessary is CalculatorPaymentSchedules, storing this should prevent having to call out to services again. The other values are used to build up user input over several pages? If my assumption is correct then I would have though they could be replaced with flash scoped variables.